### PR TITLE
Support limit in StreamingTableExec

### PR DESCRIPTION
**Pull request by lewiszlw**
## Which issue does this PR close?

Closes todo.

### CODE REVIEW

File: datafusion/physical-plan/src/streaming.rs [hunk 1]
```diff
@@ -58,7 +60,9 @@ pub struct StreamingTableExec {
     projected_schema: SchemaRef,
     projected_output_ordering: Vec<LexOrdering>,
     infinite: bool,
+    limit: Option<usize>,
```
Comment (alamb) [inline datafusion/physical-plan/src/streaming.rs:63]: Can we please also add `limit` to the display / explain plan? 

https://github.com/alamb/datafusion/blob/b8d7498da504956ee62fb0b3722f7c46e0831466/datafusion/physical-plan/src/streaming.rs#L171-L201

**Review by alamb — COMMENTED** [review]
Thank you for this work @lewiszlw  -- very nice

I think this change needs some test coverage to ensure we don't break the feature in the future due to some refactor

Perhaps we can add a new test in  

https://github.com/apache/datafusion/blob/03b314c2ee540846638d24496376489d23890ccc/datafusion/sqllogictest/test_files/filter_without_sort_exec.slt#L34

and include a LIMIT? 

That coupled with an `EXPLAIN` showing it is being used I think would be sufficient coverage.

File: datafusion/physical-plan/src/streaming.rs (hunk 1 above)
Comment (lewiszlw) [reply datafusion/physical-plan/src/streaming.rs:63]: Done

File: datafusion/sqllogictest/test_files/limit.slt [hunk 2]
```diff
@@ -39,6 +39,26 @@ STORED AS CSV
...
+# query
+query TT
+explain SELECT * FROM data LIMIT 10;
+----
+logical_plan
+01)Limit: skip=0, fetch=10
+02)--TableScan: data projection=[date, ticker, time], fetch=10
+physical_plan
+01)GlobalLimitExec: skip=0, fetch=10
+02)--StreamingTableExec: partition_sizes=1, projection=[date, ticker, time], infinite_source=true, fetch=10, output_ordering=[date@0 ASC NULLS LAST, ticker@1 ASC NULLS LAST, time@2 ASC NULLS LAST]
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/limit.slt]: 👍

**Review by alamb — APPROVED** [review]
Thank you @lewiszlw  -- this looks great.

 I also took the liberty of pushing a commit to this branch that refined the test to actually run the query (and show that the limit was used during runtime).

**Comment by alamb** [comment]
FYI @metesynnada

**Comment by metesynnada** [comment]
@alamb, this looks good to me. However, I'm concerned there might be redundant `LimitStream` overhead when both `GlobalLimitExec` and `StreamTableExec` are present in the plan, as `LimitStream` appears to be used twice.

Additionally, it would be better to use unit tests addition to `slt` tests, given that `GlobalLimitExec` already ensures the tests function properly even if `StreamTableExec` fails.

**Comment by alamb** [comment]
> @alamb, this looks good to me. However, I'm concerned there might be redundant `LimitStream` overhead when both `GlobalLimitExec` and `StreamTableExec` are present in the plan, as `LimitStream` appears to be used twice.
> 
> Additionally, it would be better to use unit tests addition to `slt` tests, given that `GlobalLimitExec` already ensures the tests function properly even if `StreamTableExec` fails.

Those are excellent points. @lewiszlw  are you able to add the unit tests? Otherwise I can try and find some time to do so

**Comment by lewiszlw** [comment]
Hmm, I don't get the point. How should I add unit tests?

**Comment by metesynnada** [comment]
> Hmm, I don't get the point. How should I add unit tests?

The executor should be capable of running independently within a test. For more insight, refer to the hash join tests in datafusion/physical-plan/src/joins/hash_join.rs. Your tests ofc will be much more simple.

Currently, a GlobalLimitExec is already implemented at the top of the tests, which means these tests wouldn't fail if the limit implementation in StreamingTableExec were to change."

**Comment by alamb** [comment]
I think the idea would be to create a `StreamingTableExec` directly and then read data from it, ensuring that the limit was applied

I didn't see any existing unit tests for `StreamingTableExec` so we would have to write some I think

https://github.com/apache/datafusion/blob/fa31c781d2cb0bdfba06dcc07bc75d9a5f9686b2/datafusion/physical-plan/src/streaming.rs#L221

Maybe I missed it

**Comment by alamb** [comment]
> > If you don't have time to do this @lewiszlw I can try and help, or perhaps we can file a ticket and do it as a follow on PR
> 
> Yeah, I'm on vacation. So might not action in time.

No problem -- I am giving it a shot

**Comment by alamb** [comment]
I pushed a unit test in [30bf3bf](https://github.com/apache/datafusion/pull/10309/commits/30bf3bff74a6aa5a15da0fecf835effbac02231b)

**Comment by alamb** [comment]
Thanks again @lewiszlw  and @metesynnada

**Pull request merged**